File: leetcode/40.py
from typing import List

class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:

        candidates.sort()
        combinations = []

        def bt(
            curr_combo: list[int],
            i: int,
            curr_sum: int
        ) -> None:

             
            if curr_sum == target:
                combinations.append(curr_combo[:])
                return

            if i == len(candidates) or curr_sum > target:
                return
          
            bt(curr_combo + [candidates[i]], i + 1, curr_sum + candidates[i]) 
            
            while i < len(candidates) - 1 and candidates[i] == candidates[i + 1]:
                i += 1
            
            bt(curr_combo, i + 1, curr_sum) 
            return

        bt([], 0, 0)
        return combinations

print(Solution().combinationSum2([2, 1, 5, 2, 2], 5))
# Should be [[1, 2, 2], [5]]

File: neetcode150/java/40.java
import java.util.ArrayList;
import java.util.List;
import java.util.Arrays;

class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        

        Arrays.sort(candidates);
        
        final List<List<Integer>> result = new ArrayList<>();
        
        // Recursively find the unique combinations that add up to the target
        class Recurse {
            void recurse(int i, int currentSum, ArrayList currentCombination) {
                // Terminal success state
                if (currentSum == target) {
                    result.add(new ArrayList(currentCombination));             
                    return;
                }

                // Terminal failure state
                if (i == candidates.length || currentSum > target) {
                    return;
                }
                
                // Include the current element and potentially more
                currentCombination.add(candidates[i]);
                recurse(i + 1, currentSum + candidates[i], currentCombination);
                currentCombination.remove(currentCombination.size() - 1);

                while (i < candidates.length - 1 && candidates[i] == candidates[i + 1]) {
                    i += 1;
                }

                recurse(i + 1, currentSum, currentCombination);
                return;
            }
        }

        Recurse recursion = new Recurse();
        recursion.recurse(0, 0, new ArrayList());
        return result;
    }
}